### PR TITLE
Removed heap allocation of AsyncAdapter R/W handlers (fix: #146)

### PR DESCRIPTION
**Pull request by MaxMaeder**
Implemented read/write handlers in `AsyncAdapter` using the Reactor pattern (following from a18cfe8 and 08d3705), replacing `getReadHandler` and `getWriteHandler` which allocated the handler closure on the heap.

Sorry this took a bit longer than I estimated–I've been prepping for the Talos interview process (I have my final one tomorrow, wish me luck!)

### CODE REVIEW

File: async_adapter.go [hunk 1]
```diff
@@ -103,17 +167,17 @@ func (a *AsyncAdapter) asyncReadNow(b []byte, readBytes int, readAll bool, cb As
 		return
 	}
 
-	a.scheduleRead(b, readBytes, readAll, cb)
+	a.scheduleRead(b, readBytes, cb)
 }
 
-func (a *AsyncAdapter) scheduleRead(b []byte, readBytes int, readAll bool, cb AsyncCallback) {
+func (a *AsyncAdapter) scheduleRead(b []byte, readBytes int, cb AsyncCallback) {
```
Comment (sergiu128) [inline async_adapter.go:173]: We can get rid of the `b []byte` param here, it's not used

File: async_adapter.go [hunk 2]
```diff
@@ -166,17 +220,17 @@ func (a *AsyncAdapter) asyncWriteNow(b []byte, writtenBytes int, writeAll bool,
 		return
 	}
 
-	a.scheduleWrite(b, writtenBytes, writeAll, cb)
+	a.scheduleWrite(b, writtenBytes, cb)
 }
 
-func (a *AsyncAdapter) scheduleWrite(b []byte, writtenBytes int, writeAll bool, cb AsyncCallback) {
+func (a *AsyncAdapter) scheduleWrite(b []byte, writtenBytes int, cb AsyncCallback) {
```
Comment (sergiu128) [inline async_adapter.go:226]: We can get rid of the `b []byte` param here, it's not used

File: async_adapter.go [hunk 3]
```diff
@@ -149,7 +202,8 @@ func (a *AsyncAdapter) AsyncWrite(b []byte, cb AsyncCallback) {
 //   - the provided buffer has been fully written after zero or several underlying
 //     write(...) operations.
 func (a *AsyncAdapter) AsyncWriteAll(b []byte, cb AsyncCallback) {
-	a.scheduleWrite(b, 0, true, cb)
+	a.writeReactor.init(b, true, cb)
+	a.scheduleWrite(b, 0, cb)
 }
 
 func (a *AsyncAdapter) asyncWriteNow(b []byte, writtenBytes int, writeAll bool, cb AsyncCallback) {
```
Comment (sergiu128) [inline async_adapter.go:209]: nitpick: let's get the naming in line with file.go. `writtenBytes -> wroteSoFar` and `readBytes -> readSoFar`.

**Review by sergiu128 — CHANGES_REQUESTED** [review]
Looks good and change is correct! Just a few notes, then I will merge. Thank you very much!

If anything is unclear about the code or you have more questions then feel free to reach out at sergiu@tungsten.computer.

**Comment by MaxMaeder** [comment]
I've made the requested changes!

Issue #150 looks interesting, I enjoy working on low-level communication and memory management stuff. I have finals coming up for school, so I probably won't get to it for at least a week.

**Review by sergiu128 — APPROVED** [review]

**Pull request merged**